File: app/src/main/java/io/github/v2compose/network/V2exService.java
package io.github.v2compose.network;

import androidx.annotation.Nullable;

import com.google.gson.FieldNamingPolicy;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import java.io.IOException;
import java.util.List;
import java.util.concurrent.TimeUnit;

import io.github.v2compose.BuildConfig;
import io.github.v2compose.util.Check;
import io.github.v2compose.util.L;
import me.ghui.fruit.Fruit;
import me.ghui.fruit.converter.retrofit.FruitConverterFactory;
import me.ghui.retrofit.converter.GlobalConverterFactory;
import me.ghui.retrofit.converter.annotations.Html;
import me.ghui.retrofit.converter.annotations.Json;
import okhttp3.HttpUrl;
import okhttp3.Interceptor;
import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Protocol;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.ResponseBody;
import okhttp3.logging.HttpLoggingInterceptor;
import okio.BufferedSource;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;


/**
 * Created by ghui on 25/03/2017.
 */

public class V2exService {
    public static final String WAP_USER_AGENT = "Mozilla/5.0 (Linux; Android 9.0; V2er Build/OPD3.170816.012) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/81.0.4044.138 Mobile Safari/537.36";
    public static final String WEB_USER_AGENT = "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_4; V2er) " +
            "AppleWebKit/537.36 (KHTML, like Gecko) Chrome/60.0.3112.90 Safari/537.36";
    public static final String UA_KEY = "user-agent";

    public static final long TIMEOUT_LENGTH = 10;
    private static V2exApi mAPI_SERVICE;
    private static Gson sGson;
    private static Fruit sFruit;
    private static WebkitCookieManagerProxy sCookieJar;
    private static OkHttpClient sHttpClient;


    public static void init() {
        if (mAPI_SERVICE == null) {
            Retrofit retrofit = new Retrofit.Builder()
                    .client(httpClient())
                    .addConverterFactory(GlobalConverterFactory
                            .create()
                            .add(FruitConverterFactory.create(fruit()), Html.class)
                            .add(GsonConverterFactory.create(gson()), Json.class))
                    .baseUrl(NetConstants.BASE_URL)
                    .build();
            mAPI_SERVICE = retrofit.create(V2exApi.class);
        }
    }

    public static V2exApi api() {
        return mAPI_SERVICE;
    }

    public static Gson gson() {
        if (sGson == null) {
            sGson = new GsonBuilder()
                    .setFieldNamingPolicy(FieldNamingPolicy.LOWER_CASE_WITH_UNDERSCORES)
                    .create();
        }
        return sGson;
    }

    public static Fruit fruit() {
        if (sFruit == null) {
            sFruit = new Fruit();
        }
        return sFruit;
    }

    public static OkHttpClient httpClient() {
        if (sHttpClient == null) {
            OkHttpClient.Builder builder = new OkHttpClient.Builder()
                    .connectTimeout(TIMEOUT_LENGTH, TimeUnit.SECONDS)
                    .cookieJar(cookieJar())
                    .retryOnConnectionFailure(true)
                    .addInterceptor(new ConfigInterceptor());
            if (BuildConfig.DEBUG) {
                builder.addInterceptor(new HttpLoggingInterceptor(L::v)
                        .setLevel(HttpLoggingInterceptor.Level.HEADERS));
            }
            sHttpClient = builder.build();
        }
        return sHttpClient;
    }

    public static WebkitCookieManagerProxy cookieJar() {
        if (sCookieJar == null) {
            sCookieJar = new WebkitCookieManagerProxy();
        }
        return sCookieJar;
    }

    private static class ConfigInterceptor implements Interceptor {
        @Override
        public Response intercept(Chain chain) throws IOException {
            Request request = chain.request();
            String ua = request.header(UA_KEY);
            if (Check.isEmpty(ua)) {
                request = request.newBuilder()
                        .addHeader("user-agent", WAP_USER_AGENT)
                        .build();
            }
            try {
                if (request.url().host().startsWith(".")) {
                    try {
                        HttpUrl.Builder httpUrlBuilder = request.url().newBuilder()
                                .host(NetConstants.WWW_HOST_NAME)
                                .setEncodedPathSegment(0, "t");
                        List<String> encodedPathSegments = request.url().encodedPathSegments();
                        for (int i = 0; i < request.url().encodedPathSegments().size(); i++) {
                            if (i < encodedPathSegments.size() - 1) {
                                httpUrlBuilder.setEncodedPathSegment(i + 1, encodedPathSegments.get(i));
                            } else {
                                httpUrlBuilder.addEncodedPathSegment(encodedPathSegments.get(i));
                            }
                        }
                        request = request.newBuilder()
                                .url(httpUrlBuilder.build())
                                .build();
                    } catch (Exception e) {
                        e.printStackTrace();
                    }
                }
                return chain.proceed(request);
            } catch (Exception e) {
                e.printStackTrace();
                return new Response.Builder()
                        .protocol(Protocol.HTTP_1_1)
                        .code(404)
                        .message("Exeception when execute chain.proceed request")
                        .body(new ResponseBody() {
                            @Nullable
                            @Override
                            public MediaType contentType() {
                                return null;
                            }

                            @Override
                            public long contentLength() {
                                return 0;
                            }

                            @Override
                            public BufferedSource source() {
                                return null;
                            }
                        })
                        .request(request).build();
            }
        }
    }

}

File: app/src/main/java/io/github/v2compose/network/V2exService.kt
package io.github.v2compose.network

import com.google.gson.Gson
import io.github.v2compose.network.bean.*
import me.ghui.fruit.Fruit
import me.ghui.fruit.converter.retrofit.FruitConverterFactory
import me.ghui.retrofit.converter.GlobalConverterFactory
import me.ghui.retrofit.converter.annotations.Html
import me.ghui.retrofit.converter.annotations.Json
import okhttp3.OkHttpClient
import okhttp3.ResponseBody
import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import retrofit2.http.*

/**
 * Created by ghui on 05/05/2017.
 */
interface V2exService {

    companion object {
        fun createV2exService(httpClient: OkHttpClient, fruit: Fruit, gson: Gson): V2exService {
            val retrofit = Retrofit.Builder()
                .client(httpClient)
                .addConverterFactory(
                    GlobalConverterFactory.create()
                        .add(FruitConverterFactory.create(fruit), Html::class.java)
                        .add(GsonConverterFactory.create(gson), Json::class.java)
                )
                .baseUrl(NetConstants.BASE_URL)
                .build()
            return retrofit.create(V2exService::class.java)
        }
    }

    @Json
    @GET("/api/topics/hot.json")
    suspend fun dailyHot(): DailyHotInfo

    @Json
    @GET("/api/nodes/show.json")
    suspend fun nodeInfo(@Query("name") name: String): NodeInfo

    @Json
    @GET("/api/nodes/s2.json")
    suspend fun nodes(): NodesInfo

    @Json
    @GET("/api/nodes/all.json")
    suspend fun allNodes(): List<Node>

    //https://www.v2ex.com/api/nodes/list.json?fields=name,title,topics,aliases&sort_by=topics&reverse=1
    @Json
    @GET("api/nodes/list.json?fields=name,title,topics,aliases&sort_by=topics&reverse=1")
    suspend fun topicNodes(): List<TopicNode>

    @Json
    @GET("/api/members/show.json")
    suspend fun userInfo(@Query("username") username: String): UserInfo

    @Json
    @GET("https://www.sov2ex.com/api/search")
    suspend fun search(
        @Query("q") keyword: String,
        @Query("from") from: Int,
        @Query("size") size: Int,
    ): SoV2EXSearchResultInfo

    // Below is html API
    @Html
    @GET("/")
    suspend fun homeNews(@Query("tab") tab: String): NewsInfo

    @Html
    @GET("/signin")
    suspend fun loginParam(): LoginParam

    @Html
    @FormUrlEncoded
    @Headers("Referer: " + NetConstants.BASE_URL + "/signin")
    @POST("/signin")
    suspend fun login(@FieldMap loginParams: Map<String, String>): LoginParam

    @Html
    @GET("/t/{id}")
    suspend fun topicDetails(@Path("id") topicId: String, @Query("p") page: Int): TopicInfo

    @Html
    @GET("/notifications")
    @Headers("user-agent: " + NetConstants.webUserAgent)
    suspend fun notifications(@Query("p") page: Int): NotificationInfo

    @Html
    @GET("/my/following")
    suspend fun myFollowingInfo(@Query("p") page: Int): MyFollowingInfo

    @Html
    @GET("/my/topics")
    suspend fun myTopicsInfo(@Query("p") page: Int): MyTopicsInfo

    @Html
    @GET("/my/nodes")
    suspend fun myNodesInfo(): MyNodesInfo

    @Html
    @GET("/")
    suspend fun nodesNavInfo(): NodesNavInfo

    @Html
    @GET("/go/{node}")
    suspend fun nodesInfo(@Path("node") node: String, @Query("p") page: Int): NodeTopicInfo

    @Html
    @GET("/member/{user}")
    suspend fun userPageInfo(@Path("user") username: String): UserPageInfo

    @Html
    @GET("/member/{user}/topics")
    suspend fun userTopics(@Path("user") username: String, @Query("p") page: Int): UserTopics

    @Html
    @GET("/member/{user}/replies")
    suspend fun userReplies(@Path("user") username: String, @Query("p") page: Int): UserReplies

    @Html
    @GET
    suspend fun bingSearch(@Url url: String): BingSearchResultInfo

    @Html
    @GET("/write")
    suspend fun createTopicPageInfo(): CreateTopicPageInfo

    @Html
    @FormUrlEncoded
    @POST("/write")
    suspend fun createTopic(@FieldMap postParams: Map<String, String>): CreateTopicPageInfo

    @Html
    @GET("/append/topic/{id}")
    suspend fun appendTopicPageInfo(
        @Header("Referer") referer: String,
        @Path("id") topicID: String
    ): AppendTopicPageInfo

    @Html
    @FormUrlEncoded
    @POST("/append/topic/{id}")
    suspend fun appendTopic(
        @Path("id") topicId: String,
        @FieldMap postParams: Map<String, String>
    ): AppendTopicPageInfo

    @Html
    @POST("/ajax/money")
    suspend fun thxMoney(): ThxResponseInfo

    @Json
    @GET("/{action}/topic/{id}")
    suspend fun getTopicAction(
        @Header("referer") referer: String,
        @Path("action") action: String,
        @Path("id") topicId: String,
        @Query("once") once: String
    ): V2exResult

    @Json
    @POST("/{action}/topic/{id}")
    suspend fun postTopicAction(
        @Header("referer") referer: String,
        @Path("action") action: String,
        @Path("id") topicId: String,
        @Query("once") once: String
    ): V2exResult

    @Json
    @GET("/{action}/reply/{id}")
    suspend fun getReplyAction(
        @Header("referer") referer: String,
        @Path("action") action: String,
        @Path("id") replyId: String,
        @Query("once") once: String,
    ): V2exResult

    @Json
    @POST("/{action}/reply/{id}")
    suspend fun postReplyAction(
        @Header("referer") referer: String,
        @Path("action") action: String,
        @Path("id") replyId: String,
        @Query("once") once: String,
    ): V2exResult

    @Json
    @POST("/ignore/reply/{id}")
    suspend fun ignoreReply(
        @Header("referer") referer: String,
        @Path("id") replyId: String,
        @Query("once") once: String,
    ): Response<ResponseBody>

    @Html
    @FormUrlEncoded
    @POST("/t/{id}")
    suspend fun replyTopic(
        @Path("id") id: String,
        @FieldMap replyMap: Map<String, String>
    ): ReplyTopicResultInfo

    // https://www.v2ex.com/follow/264541?once=87883
    // https://www.v2ex.com/unfollow/264541?once=86758
    @Html
    @GET
    suspend fun userAction(@Header("Referer") referer: String, @Url url: String): UserPageInfo

    //Request URL: https://www.v2ex.com/unfavorite/node/770?once=18542
    //Request URL: https://www.v2ex.com/favorite/node/770?once=18542
    @Html
    @GET
    suspend fun nodeAction(@Header("Referer") referer: String, @Url url: String): NodeTopicInfo

    @Html
    @GET("/mission/daily")
    suspend fun dailyInfo(): DailyInfo

    //    /mission/daily/redeem?once=84830
    @Html
    @Headers("Referer: " + NetConstants.BASE_URL + "/mission/daily")
    @GET("/mission/daily/redeem")
    suspend fun checkIn(@Query("once") once: String): DailyInfo

    @Html
    @GET("/2fa")
    suspend fun twoStepLogin(): TwoStepLoginInfo

    @Html
    @FormUrlEncoded
    @Headers("Referer: " + NetConstants.BASE_URL)
    @POST("/2fa")
    suspend fun signInTwoStep(@FieldMap map: Map<String, String>): TwoStepLoginInfo

    @Html
    @GET("/member/{user}")
    suspend fun homePageInfo(@Path("user") username: String): HomePageInfo

}